lib/three_scale/middleware/multitenant.rb

```diff
@@ -32,66 +32,58 @@ def initialize(attribute, app, env)
 
         def verify!(object)
 
-          # this controller shouldnt be checked
-          return true if @env["action_controller.request.path_parameters"]["controller"] == "provider/domains"
+          # this controller shouldn't be checked
+          return true if @env["action_dispatch.request.path_parameters"]["controller"] == "provider/domains"
 
+          # when the ActiveRecord object doesn't have tenant_id attribute at all
           unless object.respond_to?(attribute)
             # Multitenant.log "#{object} does not respond to: #{attribute}"
             return true
           end
 
-
+          # reload object if it was partially loaded from db without the `tenant_id` attribute
           begin
             current = object.send(attribute)
           rescue ActiveRecord::MissingAttributeError
             # Multitenant.log("#{object} is missing #{attribute}. Reloading and trying again")
-            fresh = object.class.send(:with_exclusive_scope) { object.class.find(object.id, :select => attribute) }
+            fresh = object.class.unscoped.find(object.id, :select => attribute)
             current = fresh.send(attribute)
           end
 
-          return true if current.nil?
+          # this is when tenant_id is not set because of a bug or in older installations the master account has it nil
+          return if current.nil?
+
+          # in newer installations master account has a tenant_id same as its id, like other providers
+          return if object.is_a?(::Account) && object.master
 
+          # once initialized a legitimate AR object with a tenant_id, all others in the request should have the same
           @original ||= current
 
           return if current == original
 
-          # we still need to check if it's master before raising a tenant leak
-          @cookie_store ||= find_cookie_store(@app)
-          @session ||= @cookie_store.send(:load_session, @env)
-
-          @master ||= Account.find_by_sql(["SELECT * FROM accounts WHERE master = ?", true]).first
+          # we still need to check if it wasn't master before raising a tenant leak
+          @master ||= ::Account.unscoped.master
 
-          if (user_id = @session.last[:user_id].presence)
-            @users ||= {}
-            @users[user_id] ||= User.find_by_sql(["SELECT * FROM users WHERE id = ? AND account_id = ?", user_id, @master.id]).present?
-            return if @users[user_id]
-          end
+          # this is supposed to match how we get the user_session in app/lib/authenticated_system/request.rb
+          # on API calls cookies are not present though, so we need to use safe navigation
+          @user_session ||= UserSession.authenticate(@env['action_dispatch.cookies']&.signed&.public_send(:[], :user_session))
```

Comment:
Interesting question here is why originally we kept `user_id` in a hash. As if we may observe multiple user_ids within the same request. Which doesn't make sense to me. Any ideas?

Reply:
No idea. It must be because they somehow expected different users querying in the same thread, and wanted to verify all of them belong to master.

Reply:
I don't see how this code could have ever worked with an assumption of multiple requests using the same checker.